v2/bundle/x509bundle/bundle.go

```diff
@@ -1,65 +1,182 @@
 package x509bundle
 
 import (
+	"bytes"
 	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
 	"io"
+	"io/ioutil"
+	"reflect"
+	"sync"
 
 	"github.com/spiffe/go-spiffe/v2/spiffeid"
 )
 
+const certType string = "CERTIFICATE"
+
 // Bundle is a collection of trusted public key material for a trust domain.
 type Bundle struct {
+	tdMtx       *sync.RWMutex
+	trustDomain spiffeid.TrustDomain
+
+	rootsMtx *sync.RWMutex
+	roots    []*x509.Certificate
 }
 
 // New creates a new bundle
 func New(trustDomain spiffeid.TrustDomain) *Bundle {
-	panic("not implemented")
+	return &Bundle{
+		trustDomain: trustDomain,
+		tdMtx:       &sync.RWMutex{},
+		roots:       []*x509.Certificate{},
+		rootsMtx:    &sync.RWMutex{},
+	}
 }
 
 // Load loads a Bundle from a file on disk.
 func Load(trustDomain spiffeid.TrustDomain, path string) (*Bundle, error) {
-	panic("not implemented")
+	fileBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open file: %v", err)
+	}
+
+	return Parse(trustDomain, fileBytes)
 }
 
 // Read decodes a bundle from a reader.
 func Read(trustDomain spiffeid.TrustDomain, r io.Reader) (*Bundle, error) {
-	panic("not implemented")
+	var b bytes.Buffer
+	_, err := b.ReadFrom(r)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read: %v", err)
+	}
+
+	return Parse(trustDomain, b.Bytes())
 }
 
 // Parse parses a bundle from bytes.
 func Parse(trustDomain spiffeid.TrustDomain, b []byte) (*Bundle, error) {
-	panic("not implemented")
+	bundle := New(trustDomain)
+	for {
+		if len(b) == 0 {
+			break
+		}
+		pemBlock, pemBytes := pem.Decode(b)
+		b = pemBytes
+		if pemBlock == nil {
+			return nil, errors.New("no PEM data found while decoding block")
+		}
+
+		if pemBlock.Type != certType {
+			return nil, fmt.Errorf(`block does not contain %q type, current type is: %q`, certType, pemBlock.Type)
+		}
+
+		cert, err := x509.ParseCertificate(pemBlock.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse certificate: %v", err)
+		}
+		bundle.AddX509Root(cert)
+	}
+	return bundle, nil
 }
 
 // TrustDomain returns the trust domain of the bundle.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	b.tdMtx.RLock()
+	defer b.tdMtx.RUnlock()
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.rootsMtx.RLock()
+	defer b.rootsMtx.RUnlock()
+	return b.roots
 }
 
 // AddX509Root adds an X.509 root to the bundle. If the root already
 // exists in the bundle, the contents of the bundle will remain unchanged.
-func (b *Bundle) AddX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) AddX509Root(root *x509.Certificate) {
+	b.rootsMtx.Lock()
+	defer b.rootsMtx.Unlock()
+
+	for _, r := range b.roots {
+		if reflect.DeepEqual(r, root) {
```

Comment:
Reflect is probably overkill here. Also, we compare certificates in a few spots... probably good to make a helper function out of it for readability.
```suggestion
		if bytes.Equal(r.Raw, root.Raw) {
```